### PR TITLE
[KYUUBI #4847] Close the session immediately when engine corrupt

### PR DESCRIPTION
**Pull request by huangzhir**
### _Why are the changes needed?_

to close https://github.com/apache/kyuubi/issues/4847

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by pan3793** [comment]
The test failure looks related, could you please fix it first? will take a detail look after dinner.

**Comment by huangzhir** [comment]
> The test failure looks related, could you please fix it first?
I have already fixed the test failure. Could you please review it?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 1]
```diff
@@ -283,4 +284,41 @@ class KyuubiSessionImpl(
       case _ => super.executeStatement(statement, confOverlay, runAsync, queryTimeout)
     }
   }
+
+  private var engineLastAlive: Long = _
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:288]: volatile?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 2]
```diff
@@ -283,4 +284,41 @@ class KyuubiSessionImpl(
...
+            + s"The time since last successful probe:"
+            + s" ${now - engineLastAlive} ms exceeds the timeout of $engineAliveTimeout ms. "
+            + s"The engine has failed $engineAliveFailCount times,"
+            + s" surpassing the maximum failure count of $engineAliveMaxFailCount.")
+          false
+        } else {
+          warn(
+            s"The engineRef[${engine.getEngineRefId}] alive probe fails," +
+              s"${now - engineLastAlive} ms exceeds timeout $engineAliveTimeout ms," +
+              s" and has failed $engineAliveFailCount times ",
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:318]: let's unify the multi-line string style: make sure space is added properly between lines and always put the space at the end of the line

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 3]
```diff
@@ -339,4 +345,24 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
...
   }
+
+  private def startEngineAliveChecker(): Unit = {
+    val interval = conf.get(KyuubiConf.ENGINE_ALIVE_PROBE_INTERVAL)
+    val checkTask = new Runnable {
+      override def run(): Unit = {
+        for (session <- allSessions) {
+          if (!session.asInstanceOf[KyuubiSessionImpl].checkEngineAlive()) {
+            try {
+              closeSession(session.handle)
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:356]: how about logging something to indicate it was closed by the checker?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 4]
```diff
@@ -339,4 +345,24 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
     Seq(userLimit, ipAddressLimit, userIpAddressLimit).find(_ > 0).map(_ =>
       SessionLimiter(userLimit, ipAddressLimit, userIpAddressLimit, userUnlimitedList.toSet))
   }
+
+  private def startEngineAliveChecker(): Unit = {
+    val interval = conf.get(KyuubiConf.ENGINE_ALIVE_PROBE_INTERVAL)
+    val checkTask = new Runnable {
+      override def run(): Unit = {
+        for (session <- allSessions) {
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:353]: The function style is preferred

```suggestion
        allSessions.foreach { session =>
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 5]
```diff
@@ -339,4 +345,24 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
     Seq(userLimit, ipAddressLimit, userIpAddressLimit).find(_ > 0).map(_ =>
       SessionLimiter(userLimit, ipAddressLimit, userIpAddressLimit, userUnlimitedList.toSet))
   }
+
+  private def startEngineAliveChecker(): Unit = {
+    val interval = conf.get(KyuubiConf.ENGINE_ALIVE_PROBE_INTERVAL)
+    val checkTask = new Runnable {
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:351]: Scala 2.12 and above support the SAM syntax.
```patch
- val checkTask = new Runnable {
-   override def run(): Unit = {
-     ...
-   }
- }
+ val task: Runnable = () =>
+   ...
+ }
```

**Comment by pan3793** [comment]
Thanks, the change makes sense to me.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 6]
```diff
@@ -339,4 +345,24 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
...
+          try {
+            closeSession(session.handle)
+            logger.info(s"The session ${session.handle} has been closed " +
+              s"due to engine unresponsiveness (checked by the engine alive checker).")
+          } catch {
+            case e: KyuubiSQLException =>
+              warn(s"Error closing session ${session.handle}", e)
+          }
+        }
+      })
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:363]:
```suggestion
      allSessions.foreach { session =>
        if (!session.asInstanceOf[KyuubiSessionImpl].checkEngineAlive()) {
          try {
            closeSession(session.handle)
            logger.info(s"The session ${session.handle} has been closed " +
              s"due to engine unresponsiveness (checked by the engine alive checker).")
          } catch {
            case e: KyuubiSQLException =>
              warn(s"Error closing session ${session.handle}", e)
          }
        }
      }
```

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master

**Comment by bowenliang123** [comment]
There's no milestone marked on this PR. Is this merged to 1.8.0 only?

**Comment by pan3793** [comment]
thanks for the top, updated, it was merged before cutting branch-1.8

**Comment by bowenliang123** [comment]
Thanks, understand. It should be helpful for production use.

**Pull request closed without merging**